### ISSUE TITLE
[cpp-clang] Broken multiple-inheritance

### ISSUE THREAD

**Issue by intrigus-lgtm** [esbmc/esbmc#1866]
esbmc (32ab7a8680c08f8172ad3a0c5796695f67dc4155) fails for this very simple code:

```
#include <assert.h>

struct B
{   
  char e = 22;
};
struct A
{
  char b = 111;
};
struct f : B, A
{   
};
int main()
{
  f i = f();
  assert(i.b == 111);
  assert(i.e == 22);
}
```
Fails with this log:
<details>

```
Generating GOTO Program
GOTO program creation time: 0.437s
GOTO program processing time: 0.000s
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

__ESBMC_pthread_start_main_hook (c:@F@__ESBMC_pthread_start_main_hook):
        // 0 no location
        FUNCTION_CALL:  yield()
        // 1 file pthread_lib.c line 118 column 3 function __ESBMC_pthread_start_main_hook
        ATOMIC_BEGIN
        // 2 file pthread_lib.c line 119 column 3 function __ESBMC_pthread_start_main_hook
        num_total_threads=num_total_threads + 1;
        // 3 file pthread_lib.c line 120 column 3 function __ESBMC_pthread_start_main_hook
        num_threads_running=num_threads_running + 1;
        // 4 file pthread_lib.c line 121 column 3 function __ESBMC_pthread_start_main_hook
        ATOMIC_END
        // 5 file pthread_lib.c line 122 column 1 function __ESBMC_pthread_start_main_hook
        END_FUNCTION // __ESBMC_pthread_start_main_hook
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

__ESBMC_pthread_end_main_hook (c:@F@__ESBMC_pthread_end_main_hook):
        // 6 no location
        FUNCTION_CALL:  yield()
        // 7 file pthread_lib.c line 130 column 3 function __ESBMC_pthread_end_main_hook
        ATOMIC_BEGIN
        // 8 file pthread_lib.c line 131 column 3 function __ESBMC_pthread_end_main_hook
        num_threads_running=num_threads_running - 1;
        // 9 file pthread_lib.c line 132 column 1 function __ESBMC_pthread_end_main_hook
        END_FUNCTION // __ESBMC_pthread_end_main_hook
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

__ESBMC_atexit_handler (c:@F@__ESBMC_atexit_handler):
        // 10 file stdlib.c line 33 column 3 function __ESBMC_atexit_handler
        // Labels: __ESBMC_HIDE
        IF atexits == (struct struct atexit_key *)0 THEN GOTO 2
        // 11 file stdlib.c line 36 column 3 function __ESBMC_atexit_handler
     1: IF !(_Bool)atexits THEN GOTO 2
        // 12 file stdlib.c line 38 column 5 function __ESBMC_atexit_handler
        FUNCTION_CALL:  *atexits->atexit_func()
        // 13 file stdlib.c line 39 column 5 function __ESBMC_atexit_handler
        struct struct atexit_key * tmp;
        // 14 file stdlib.c line 39 column 5 function __ESBMC_atexit_handler
        tmp=atexits->next;
        // 15 file stdlib.c line 40 column 5 function __ESBMC_atexit_handler
        FREE((void *)atexits);
        // 16 file stdlib.c line 41 column 5 function __ESBMC_atexit_handler
        atexits=tmp;
        // 17 file stdlib.c line 42 column 3 function __ESBMC_atexit_handler
        dead tmp;
        // 18 file stdlib.c line 36 column 3 function __ESBMC_atexit_handler
        GOTO 1
        // 19 file stdlib.c line 43 column 1 function __ESBMC_atexit_handler
     2: END_FUNCTION // __ESBMC_atexit_handler
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

B (c:@S@B@F@B#):
        // 20 no location
        THROW_DECL (noexcept)
        // 21 
        this->e=22;
        // 22 file cpptest.pre.cpp line 3 column 8 function B
        END_FUNCTION // B
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A (c:@S@A@F@A#):
        // 23 no location
        THROW_DECL (noexcept)
        // 24 
        this->b=111;
        // 25 file cpptest.pre.cpp line 7 column 8 function A
        END_FUNCTION // A
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

f (c:@S@f@F@f#):
        // 26 no location
        THROW_DECL (noexcept)
        // 27 
        FUNCTION_CALL:  B(( B *)this)
        // 28 
        FUNCTION_CALL:  A(( A *)this)
        // 29 file cpptest.pre.cpp line 11 column 8 function f
        END_FUNCTION // f
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

main (c:@F@main#):
        // 30 file cpptest.pre.cpp line 16 column 3 function main
        f i;
        // 31 file cpptest.pre.cpp line 16 column 9 function main
        f tmp$1;
        // 32 file cpptest.pre.cpp line 16 column 9 function main
        tmp$1=NONDET(struct);
        // 33 
        FUNCTION_CALL:  f(&tmp$1)
        // 34 file cpptest.pre.cpp line 16 column 3 function main
        i=tmp$1;
        // 35 file cpptest.pre.cpp line 17 column 3 function main
        ASSERT (signed int)i.b == 111 // assertion i.b == 111
        // 36 file cpptest.pre.cpp line 17 column 3 function main
        IF !((signed int)i.b == 111) THEN GOTO 1
        // 37 file cpptest.pre.cpp line 17 column 3 function main
        0;

        // 38 file cpptest.pre.cpp line 18 column 3 function main
     1: ASSERT (signed int)i.e == 22 // assertion i.e == 22
        // 39 file cpptest.pre.cpp line 18 column 3 function main
        IF !((signed int)i.e == 22) THEN GOTO 2
        // 40 file cpptest.pre.cpp line 18 column 3 function main
        0;

        // 41 no location
     2: FUNCTION_CALL:  ~f(&i)
        // 42 file cpptest.pre.cpp line 19 column 1 function main
        dead i;
        // 43 no location
        FUNCTION_CALL:  ~f(&tmp$1)
        // 44 file cpptest.pre.cpp line 19 column 1 function main
        dead tmp$1;
        // 45 file cpptest.pre.cpp line 19 column 1 function main
        RETURN: NONDET(signed int)
        // 46 file cpptest.pre.cpp line 19 column 1 function main
        END_FUNCTION // main
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

__ESBMC_main (__ESBMC_main):
        // 47 file esbmc_intrinsics.h line 16 column 1
        alloc=ARRAY_OF(0);
        // 48 file esbmc_intrinsics.h line 19 column 1
        is_dynamic=ARRAY_OF(0);
        // 49 file esbmc_intrinsics.h line 22 column 1
        alloc_size={ 0 };
        // 50 file esbmc_intrinsics.h line 30 column 1
        rounding_mode=0;
        // 51 file pthread_lib.c line 49 column 1
        num_threads_running=0;
        // 52 file stdlib.c line 27 column 1
        atexits=0;
        // 53 file pthread_lib.c line 48 column 1
        num_total_threads=0;
        // 54 no location
        FUNCTION_CALL:  pthread_start_main_hook()
        // 55 no location
        FUNCTION_CALL:  main()
        // 56 no location
        FUNCTION_CALL:  atexit_handler()
        // 57 no location
        FUNCTION_CALL:  pthread_end_main_hook()
        // 58 file cpptest.pre.cpp line 19 column 1 function main
        END_FUNCTION // main

Starting Bounded Model Checking
WARNING: no body for function ~f#
WARNING: no body for function ~f#
Symex completed in: 0.003s (22 assignments)
Slicing time: 0.000s (removed 3 assignments)

Program constraints: 
// 47 file esbmc_intrinsics.h line 16 column 1
/* 1 */ alloc&0#1 == ARRAY_OF(0)

// 48 file esbmc_intrinsics.h line 19 column 1
/* 2 */ is_dynamic&0#1 == ARRAY_OF(0)

// 49 file esbmc_intrinsics.h line 22 column 1
/* 3 */ alloc_size&0#1 == { 0 }

// 50 file esbmc_intrinsics.h line 30 column 1
/* 4 */ rounding_mode&0#1 == 0

// 51 file pthread_lib.c line 49 column 1
/* 5 */ num_threads_running&0#1 == 0

// 52 file stdlib.c line 27 column 1
/* 6 */ atexits&0#1 == 0

// 53 file pthread_lib.c line 48 column 1
/* 7 */ num_total_threads&0#1 == 0

// 2 file pthread_lib.c line 119 column 3 function __ESBMC_pthread_start_main_hook
/* 8 */ num_total_threads&0#2 == 1

// 3 file pthread_lib.c line 120 column 3 function __ESBMC_pthread_start_main_hook
/* 9 */ num_threads_running&0#2 == 1

// 32 file cpptest.pre.cpp line 16 column 9 function main
/* 10 */ main#::$tmp::tmp$1?1!0&0#1 == nondet_symbol(symex::nondet0)

// 33 
/* 11 */ f#this?1!0&0#1 == &main#::$tmp::tmp$1?1!0

// 27 
/* 12 */ B#this?1!0&0#1 == ( B *)(&main#::$tmp::tmp$1?1!0)

// 21 
/* 13 */ main#::$tmp::tmp$1?1!0&0#2 == (main#::$tmp::tmp$1?1!0&0#1 WITH [b:=22])

// 28 
/* 14 */ A#this?1!0&0#1 == ( A *)(&main#::$tmp::tmp$1?1!0)

// 24 
/* 15 */ main#::$tmp::tmp$1?1!0&0#3 == (main#::$tmp::tmp$1?1!0&0#2 WITH [b:=111])

// 34 file cpptest.pre.cpp line 16 column 3 function main
/* 16 */ i?1!0&0#1 == main#::$tmp::tmp$1?1!0&0#3

// 35 file cpptest.pre.cpp line 17 column 3 function main (assertion i.b == 111)
/* 17 */ (assert)execution_statet::\guard_exec?0!0 => (signed int)i?1!0&0#1.b == 111

// 37 file cpptest.pre.cpp line 17 column 3 function main
/* 18 */ goto_symex::guard?0!0&0#1 == ((signed int)i?1!0&0#1.b == 111)

// 38 file cpptest.pre.cpp line 18 column 3 function main (assertion i.e == 22)
/* 19 */ (assert)execution_statet::\guard_exec?0!0 => (signed int)i?1!0&0#1.e == 22


Generated 6 VCC(s), 2 remaining after simplification (19 assignments)
No solver specified; defaulting to z3
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Encoding to solver time: 0.000s
Total number of safety properties: 34
Solving with solver Z3 v4.8.12
Runtime decision procedure: 0.002s
(define-fun |c:@F@main#::$tmp::tmp$1?1!0&0#1| () struct_type_f
  (struct_type_f #x00 #x80))
(define-fun __ESBMC_ptr_obj_end_2 () (_ BitVec 64)
  #x8000000000000000)
(define-fun |execution_statet::\\guard_exec?0!0| () Bool
  true)
(define-fun |goto_symex::guard?0!0&0#1| () Bool
  true)
(define-fun |c:cpptest.pre.cpp@122@F@main#@i?1!0&0#1| () struct_type_f
  (struct_type_f #x6f #x80))
(define-fun |c:@F@main#::$tmp::tmp$1?1!0&0#3| () struct_type_f
  (struct_type_f #x6f #x80))
(define-fun |c:@S@A@F@A#this?1!0&0#1| () struct_type_pointer_struct
  (struct_type_pointer_struct #x0000000000000002 #x0000000000000000))
(define-fun |c:@F@main#::$tmp::tmp$1?1!0&0#2| () struct_type_f
  (struct_type_f #x16 #x80))
(define-fun |c:@S@B@F@B#this?1!0&0#1| () struct_type_pointer_struct
  (struct_type_pointer_struct #x0000000000000002 #x0000000000000000))
(define-fun |c:@S@f@F@f#this?1!0&0#1| () struct_type_pointer_struct
  (struct_type_pointer_struct #x0000000000000002 #x0000000000000000))
(define-fun |c:@F@main#::$tmp::tmp$1?1!0| () struct_type_pointer_struct
  (struct_type_pointer_struct #x0000000000000002 #x0000000000000000))
(define-fun __ESBMC_addrspace_arr_1 () (Array (_ BitVec 64) struct_type_addr_space_type)
  ((as const (Array (_ BitVec 64) struct_type_addr_space_type))
  (struct_type_addr_space_type #x0000000000000000 #x0000000000000000)))
(define-fun __ESBMC_addrspace_arr_4 () (Array (_ BitVec 64) struct_type_addr_space_type)
  (store (store ((as const (Array (_ BitVec 64) struct_type_addr_space_type))
                (struct_type_addr_space_type
                  #x0000000000000000
                  #x0000000000000000))
              #x0000000000000001
              (struct_type_addr_space_type
                #x0000000000000001
                #xffffffffffffffff))
       #x0000000000000002
       (struct_type_addr_space_type #x7ffffffffffffffe #x8000000000000000)))
(define-fun __ESBMC_ptr_addr_range_2 () struct_type_addr_space_type
  (struct_type_addr_space_type #x7ffffffffffffffe #x8000000000000000))
(define-fun __ESBMC_ptr_obj_start_2 () (_ BitVec 64)
  #x7ffffffffffffffe)
(define-fun |nondet$symex::nondet0| () struct_type_f
  (struct_type_f #x00 #x80))
(define-fun |c:@__ESBMC_num_threads_running&0#2| () (_ BitVec 16)
  #x0001)
(define-fun |c:@__ESBMC_num_total_threads&0#2| () (_ BitVec 16)
  #x0001)
(define-fun |c:@__ESBMC_num_total_threads&0#1| () (_ BitVec 16)
  #x0000)
(define-fun |c:stdlib.c@__ESBMC_atexits&0#1| () struct_type_pointer_struct
  (struct_type_pointer_struct #x0000000000000000 #x0000000000000000))
(define-fun |c:@__ESBMC_num_threads_running&0#1| () (_ BitVec 16)
  #x0000)
(define-fun |c:@__ESBMC_rounding_mode&0#1| () (_ BitVec 32)
  #x00000000)
(define-fun |c:@__ESBMC_alloc_size&0#1| () (Array (_ BitVec 64) (_ BitVec 64))
  ((as const (Array (_ BitVec 64) (_ BitVec 64))) #x0000000000000000))
(define-fun |c:@__ESBMC_is_dynamic&0#1| () (Array (_ BitVec 64) Bool)
  ((as const (Array (_ BitVec 64) Bool)) false))
(define-fun |c:@__ESBMC_alloc&0#1| () (Array (_ BitVec 64) Bool)
  ((as const (Array (_ BitVec 64) Bool)) false))
(define-fun INVALID () struct_type_pointer_struct
  (struct_type_pointer_struct #x0000000000000001 #x0000000000000000))
(define-fun NULL () struct_type_pointer_struct
  (struct_type_pointer_struct #x0000000000000000 #x0000000000000000))
(define-fun __ESBMC_addrspace_arr_3 () (Array (_ BitVec 64) struct_type_addr_space_type)
  (store ((as const (Array (_ BitVec 64) struct_type_addr_space_type))
         (struct_type_addr_space_type #x0000000000000000 #x0000000000000000))
       #x0000000000000001
       (struct_type_addr_space_type #x0000000000000001 #xffffffffffffffff)))
(define-fun __ESBMC_addrspace_arr_2 () (Array (_ BitVec 64) struct_type_addr_space_type)
  ((as const (Array (_ BitVec 64) struct_type_addr_space_type))
  (struct_type_addr_space_type #x0000000000000000 #x0000000000000000)))
(define-fun __ESBMC_ptr_addr_range_1 () struct_type_addr_space_type
  (struct_type_addr_space_type #x0000000000000001 #xffffffffffffffff))
(define-fun __ESBMC_ptr_addr_range_0 () struct_type_addr_space_type
  (struct_type_addr_space_type #x0000000000000000 #x0000000000000000))
(define-fun __ESBMC_ptr_obj_end_1 () (_ BitVec 64)
  #xffffffffffffffff)
(define-fun __ESBMC_ptr_obj_start_1 () (_ BitVec 64)
  #x0000000000000001)
(define-fun __ESBMC_ptr_obj_end_0 () (_ BitVec 64)
  #x0000000000000000)
(define-fun __ESBMC_ptr_obj_start_0 () (_ BitVec 64)
  #x0000000000000000)
Building error trace

[Counterexample]


State 1 file esbmc_intrinsics.h line 16 column 1 thread 0
----------------------------------------------------
  alloc = ARRAY_OF(0)

State 2 file esbmc_intrinsics.h line 19 column 1 thread 0
----------------------------------------------------
  is_dynamic = ARRAY_OF(0)

State 3 file esbmc_intrinsics.h line 22 column 1 thread 0
----------------------------------------------------
  alloc_size = { 0 }

State 4 file esbmc_intrinsics.h line 30 column 1 thread 0
----------------------------------------------------
  rounding_mode = 0 (00000000 00000000 00000000 00000000)

State 5 file pthread_lib.c line 49 column 1 thread 0
----------------------------------------------------
  num_threads_running = 0 (00000000 00000000)

State 6 file stdlib.c line 27 column 1 thread 0
----------------------------------------------------
  atexits = 0

State 7 file pthread_lib.c line 48 column 1 thread 0
----------------------------------------------------
  num_total_threads = 0 (00000000 00000000)

State 8 file pthread_lib.c line 119 column 3 function __ESBMC_pthread_start_main_hook thread 0
----------------------------------------------------
  num_total_threads = 1 (00000000 00000001)

State 9 file pthread_lib.c line 120 column 3 function __ESBMC_pthread_start_main_hook thread 0
----------------------------------------------------
  num_threads_running = 1 (00000000 00000001)

State 10 file cpptest.pre.cpp line 16 column 9 function main thread 0
----------------------------------------------------
  tmp$1 = { .b=0, .e=-128 }

State 11  thread 0
----------------------------------------------------
  this = &tmp$1

State 12  thread 0
----------------------------------------------------
  this = ( B *)(&tmp$1)

State 13  thread 0
----------------------------------------------------
  this->e = 22 (00010110)

State 14  thread 0
----------------------------------------------------
  this = ( A *)(&tmp$1)

State 15  thread 0
----------------------------------------------------
  this->b = 111 (01101111)

State 16 file cpptest.pre.cpp line 16 column 3 function main thread 0
----------------------------------------------------
  i = { .b=111, .e=-128 }

State 19 file cpptest.pre.cpp line 18 column 3 function main thread 0
----------------------------------------------------
Violated property:
  file cpptest.pre.cpp line 18 column 3 function main
  assertion i.e == 22
  (signed int)i.e == 22


VERIFICATION FAILED

Process finished with exit code 1

```
</details>

If anybody already knows what causes this and can fix it, I'd be very happy.
I'd also be happy to take pointers in the right direction and fix this myself.
I would _hope_ that this is something simple like having to add a cast to the right sub object...

**Comment by intrigus-lgtm** [comment]
Slightly related broken test case:
```cpp
#include <assert.h>

struct B
{   
  char e = 22;
};
struct A
{
  char e = 111;
};
struct f : B, A
{   
};
int main()
{
  f i = f();
  assert(i.B::e == 22); // WORKS if changed to 111 as as well!
  assert(i.A::e == 111);
}
```

**Comment by rafaelsamenezes** [comment]
> If anybody already knows what causes this and can fix it, I'd be very happy.
I'd also be happy to take pointers in the right direction and fix this myself.
I would hope that this is something simple like having to add a cast to the right sub object...

I don't have any specifics advice rather than to do further investigation:

- Check the type and fields through the symbol table.
- Check the vtable.
- Check the parsing tree
- Check how the inheritance is taking place in the convert.

**Comment by intrigus-lgtm** [comment]
@lucasccordeiro it is my impression that multiple inheritance in esbmc is fundamentally broken (and probably has been in the new front-end since its existence).
This is because esbmc does not adjust pointers when casting.
For example, consider this program:
```cpp
#include <assert.h>
#include <stdio.h>
#include <cstddef>

struct B
{
  long f = 11;
};
struct A
{
  long f = 22;
};
struct F : A, B
{
  long f = 33;
};

F i = F();
F *ip = &i;
A *ap = ip; // *1*
B *bp = ip; // *2*
// B *betterbp = (B *)(((char *)ip) + 8);
int main()
{
  assert((void *)bp != (void *)ip);
  // ptrdiff_t offset = (char *)bp - (char *)ip;
  // assert(offset == 8);
  assert(bp->f == 11);
  assert(ap->f == 22);
  assert(ip->f == 33);
}
```

When we assign `ip` to `ap`, we cast from `F*` to `A*`.
No adjustment of the pointer is necessary, because `F` and `A` share the same offset in the object.
However, when we assign `ip` to `bp`, we have to adjust the pointer, because `F` and `B` do NOT share the same offset in the object. This adjustment would look like this: `(B *)(((char *)ip) + 8);`
esbmc is not doing this.
The result is that when we do `bp->f`, we use the offset for `f` in `B` (as one would expect!), but `bp` does NOT actually point to an object of type `B`.
I fixed the missing adjustment problem.

**However**, if we add virtual inheritance, this offset is no longer static, but is dynamic and so we would have to record that offset in the vtable or similar.
This would open a whole can of worms IMHO.

So we have as far as I understand two options:
1. Go down the rabbit hole of implementing vbase offsets and stuff like that; basically: implement the *whole* [C++ Itanium ABI](https://itanium-cxx-abi.github.io/cxx-abi/abi.html) which is insanely complex IMHO (see for example: https://itanium-cxx-abi.github.io/cxx-abi/abi.html#vtable-construction, https://itanium-cxx-abi.github.io/cxx-abi/abi.html#vtable-ctor).
While esbmc/the solvers it uses might have no problem with the pointer adjustments itself (performance-wise) (no idea if this is actually no problem), it should looks super painful to implement correctly.
2. Implement [this paper](https://link.springer.com/content/pdf/10.1007/978-3-642-28652-0_8.pdf). To me, this looks far simpler than the Itanium C++ ABI (or any other ABI).

Basically, we can either fully implement the left side or try the right side:
![grafik](https://github.com/esbmc/esbmc/assets/60750685/7b387a00-21b3-4805-a43c-2bb41f48e8d4)

**Please let me know if there is anything I'm missing or misunderstanding.**

**Comment by intrigus-lgtm** [comment]
I'm aware that esbmc does not actually implement the Itanium ABI, but something inspired by it.
Still, I would probably not be confident (which is bad considering that esbmc is a verification tool) to extend its support to virtual bases.

**Comment by lucasccordeiro** [comment]
Hi @intrigus-lgtm,

Thanks for your message.

I have asked @XLiZHI to take a closer look at this issue.

In the meantime, can I ask you whether you had a chance to check our wiki page at https://github.com/esbmc/esbmc/wiki/ESBMC-Cpp-Support?

There are still some TODOs to add full support for C++ inheritance and polymorphism in ESBMC:

````
correct order of ctors/dtors:
tracked by the umbrella issue [issue 940](https://github.com/esbmc/esbmc/issues/940), affected features are listed below:
Virtual destructor
Base initialization for the most-derivied class
Order of destruction in case of object composition (part-whole relationship)
````
